### PR TITLE
Fix SoundSourceMediaFoundation Plugin (M4A, Windows)

### PR DESCRIPTION
**Pull request by uklotzde**
My New Year gift for you: A working and sample accurate .m4a decoder for Windows.

Please test this branch extensively and send me files that cause any issues!! For this purpose it would be helpful if you run a debug version of Mixxx that has been built with the following _scons_ options:
```
scons ... qdebug=1 debug_assertions_fatal=1
```

### CODE REVIEW

**Comment by rryan** [comment]
Woohoo! Thanks

File: src/test/soundproxy_test.cpp [hunk 1]
```diff
@@ -21,15 +21,16 @@ class SoundSourceProxyTest: public MixxxTest {
...
-        // TODO(XXX): The seek tests currently fail for the
-        // SoundSourceMediaFoundation plugin on windows when
-        // decoding M4A files.
-        availableFileNameSuffixes << ".m4a";
-#endif
+        availableFileNameSuffixes
+                << ".aiff"
+                << ".flac"
+                << ".m4a"
+                << "-png.mp3"
```
Comment (daschuer) [inline src/test/soundproxy_test.cpp]: what is -png-mp3 ?

**Comment by uklotzde** [comment]
We have 2 MP3 test files, one with a .jpg and the other with .png cover art.

**Comment by uklotzde** [comment]
Superfast merge, thanks! Considering the fact that it couldn't get any worse and we might get much better testing now, your decision is reasonable 👍 I'm not using Windows myself on a daily basis.

I will certainly submit some some follow ups, e.g. minor seek optimizations I already have in mind.

File: plugins/soundsourcemediafoundation/soundsourcemediafoundation.cpp [hunk 2]
```diff
@@ -160,233 +124,273 @@ void SoundSourceMediaFoundation::close() {
...
+    if (m_pSourceReader == nullptr) {
+        // reader is dead -> jump to end of stream
+        return getMaxFrameIndex();
     }
 
-    // http://msdn.microsoft.com/en-us/library/dd374668(v=VS.85).aspx
-    hr = m_pReader->SetCurrentPosition(GUID_NULL, prop);
+    HRESULT hr;
+
+    hr = m_pSourceReader->Flush(MF_SOURCE_READER_FIRST_AUDIO_STREAM);
```
Comment (daschuer) [inline plugins/soundsourcemediafoundation/soundsourcemediafoundation.cpp]: Here a comment can be handy, please improve: 
// Here we need to either jump far ahead of backward so the stored frames in th reader a useless.

File: plugins/soundsourcemediafoundation/soundsourcemediafoundation.cpp [hunk 3]
```diff
@@ -160,233 +124,273 @@ void SoundSourceMediaFoundation::close() {
...
+        qWarning() << "SoundSourceMediaFoundation:"
+                << "Failed to flush reader before seeking"
+                << hr;
+    }
+    // Flush left over buffer
+    m_sampleBufferCount = 0;
+    m_sampleBufferOffset = 0;
+
+    // Invalidate current position
+    m_currentFrameIndex = getMaxFrameIndex(); // invalidate
```
Comment (daschuer) [inline plugins/soundsourcemediafoundation/soundsourcemediafoundation.cpp]: redundant comment

File: plugins/soundsourcemediafoundation/soundsourcemediafoundation.cpp [hunk 4]
```diff
@@ -160,233 +124,273 @@ void SoundSourceMediaFoundation::close() {
...
+    DEBUG_ASSERT(SUCCEEDED(hr)); // never fails
+    hr = m_pSourceReader->SetCurrentPosition(GUID_NULL, prop);
+    PropVariantClear(&prop);
+    if (SUCCEEDED(hr)) {
+        // Restart decoding by skipping over the prefetch samples
+        DEBUG_ASSERT(seekIndex < frameIndex);
+        SINT skipFramesCount = frameIndex - seekIndex;
+        // NOTE(uklotzde): This loop is needed, because often the
+        // target position has not been reached after the first
+        // invocation of skipSampleFrames()!?!?
```
Comment (daschuer) [inline plugins/soundsourcemediafoundation/soundsourcemediafoundation.cpp]: Why !?!? is this a assumption we may verify?

File: plugins/soundsourcemediafoundation/soundsourcemediafoundation.cpp [hunk 5]
```diff
@@ -160,233 +124,273 @@ void SoundSourceMediaFoundation::close() {
...
-                qWarning() << __FILE__ << __LINE__
-                        << "WARNING: Expected frames needed to be 0. Abandoning this file.";
-                m_dead = true;
+
+    SINT numberOfFramesRemaining = numberOfFrames;
+    CSAMPLE* pSampleBuffer = sampleBuffer;
+
+    while (numberOfFramesRemaining > 0) {
+        SINT copySamplesCount =
+            std::min(frames2samples(numberOfFramesRemaining), m_sampleBufferCount);
```
Comment (daschuer) [inline plugins/soundsourcemediafoundation/soundsourcemediafoundation.cpp]: sample vs frames: Should it be a "m_frameBufferCount"

Comment (daschuer) [inline plugins/soundsourcemediafoundation/soundsourcemediafoundation.cpp]: or m_sampleBufferFrames

Comment (daschuer) [inline plugins/soundsourcemediafoundation/soundsourcemediafoundation.cpp]: Oh sorry for the noise. Now I understand everything is fine.

File: plugins/soundsourcemediafoundation/soundsourcemediafoundation.cpp [hunk 6]
```diff
@@ -160,233 +124,273 @@ void SoundSourceMediaFoundation::close() {
...
+
+        if (m_pSourceReader == nullptr) {
+            break; // abort (reader is dead)
         }
 
+        DWORD dwFlags = 0;
+        qint64 streamPos = 0;
+        IMFSample* pSample = nullptr;
+        HRESULT hr = m_pSourceReader->ReadSample(
+            MF_SOURCE_READER_FIRST_AUDIO_STREAM, // [in] DWORD dwStreamIndex,
```
Comment (daschuer) [inline plugins/soundsourcemediafoundation/soundsourcemediafoundation.cpp]: An extra indent to distinguish this from a logical block is missing here.

**Comment by uklotzde** [comment]
I've already started with the follow-up work and noticed this, too ;)

**Comment by uklotzde** [comment]
This is the only issue that I still don't understand completely. The first 
attempt to read samples from the reader obviously starts at a position 
before the position we have explicitly set. So we need to call 
skipSampleFrames() twice, 1st starting at an unknown/invalid position 
followed by a 2nd call from a known position.

**Comment by uklotzde** [comment]
Damn, I forgot this one -> samples2frames()! The sample buffer stores 
individual samples (from complete frames, of course)

**Comment by uklotzde** [comment]
I didn't spot the inverse frames2samples() call of the 1st argument either 
:D Yes, it's correct.

**Comment by daschuer** [comment]
LGTM thank you.

**Comment by uklotzde** [comment]
The flush should be unnecssary when using synchronous calls. I will test 
this later.

**Pull request merged**